plugins/modules/firmware_update.py

```diff
@@ -0,0 +1,316 @@
+    try:
+        initial_level = hmc.get_firmware_level(system_name)
+        hmc.update_managed_system(system_name, False, repo, level, remote_repo)
+        ret_dict = {'msg': 'system update finished'}
+        new_level = hmc.get_firmware_level(system_name)
+        logger.debug("new_level: %s", new_level)
+        ret_dict.update(new_level)
+    except HmcError as on_system_error:
+        return False, None, repr(on_system_error)
+
+    changed = True
+    ret_dict['diff'] = {'before': initial_level,
+                        'after': new_level,
+                        }
+    return changed, ret_dict, None
+
+
+def upgrade_system(module, params):
+    hmc = create_hmc_conn(module, params)
+    system_name, repo, level, remote_repo = extract_updlic_options(params)
+    ret_dict = {}
+    try:
+        initial_level = hmc.get_firmware_level(system_name)
+        hmc.update_managed_system(system_name, True, repo, level, remote_repo)
+        ret_dict = {'msg': 'system upgrade finished'}
+        new_level = hmc.get_firmware_level(system_name)
+        logger.debug("new_level: %s", new_level)
+        ret_dict.update(new_level)
+    except HmcError as on_system_error:
+        return False, None, repr(on_system_error)
+
+    changed = True
+    ret_dict['diff'] = {'before': initial_level,
+                        'after': new_level,
+                        }
+    return changed, ret_dict, None
+
+
+def accept_level(module, params):
+    hmc = create_hmc_conn(module, params)
+    system_name = params['system_name']
+    ret_dict = {}
+    try:
+        hmc.accept_level(system_name)
+    except HmcError as on_system_error:
+        return False, None, repr(on_system_error)
+
+    ret_dict['msg'] = 'level accepted'
+    return True, ret_dict, None
+
+
+def perform_task(module):
+    params = module.params
+    actions = {
+        "updated": update_system,
+        "upgraded": upgrade_system,
+        "accept": accept_level,
+    }
+    oper = 'action'
+    if params['action'] is None:
+        oper = 'state'
+    try:
+        return actions[params[oper]](module, params)
+    except Exception as error:
+        return False, repr(error), None
+
+
+def run_module():
+    module_args = dict(
+        hmc_host=dict(type='str', required=True),
+        hmc_auth=dict(type='dict',
+                      required=True,
+                      no_log=True,
+                      options=dict(
+                          username=dict(required=True, type='str'),
+                          password=dict(type='str', no_log=True),
+                      )
+                      ),
+        system_name=dict(type='str', required=True),
+        action=dict(type='str', choices=['accept']),
+        state=dict(type='str', choices=['updated', 'upgraded', ]),
+        level=dict(type='str', default='latest'),
+        repository=dict(type='str', default='ibmwebsite', choices=['ibmwebsite', 'ftp', 'sftp']),
+        remote_repo=dict(type='dict', options=dict(
```

Comment:
I think this needs to be made mandatory when the state is set to updated or upgraded

Reply:
The reason it's not mandatory is because the ibmwebsite choice doesn't support the remote_repo options on updlic. So there can be update/upgrade calls that don't have any remote_repo locations.